File: cpp/src/examples/basic_encrypt.py
# Encrypts and decrypts a short message. Uses raw encoding, base64w
# encoding and zlib compression.
#
# Example: python basic_encrypt.py ~/my-aes
#
import os
import sys

import keyczar

def EncryptAndDecrypt(keyset_path):
    input = 'Secret message'
    encrypter = keyczar.Encrypter.Read(keyset_path)
    ciphertext_b64 = encrypter.Encrypt(input)

    print 'plaintext:', input
    print 'ciphertext (base64w):', ciphertext_b64

    crypter = keyczar.Crypter.Read(keyset_path)
    assert crypter.Decrypt(ciphertext_b64) == input

def EncryptAndDecryptBytes(keyset_path):
    input = 'Secret message'
    crypter = keyczar.Crypter.Read(keyset_path)
    crypter.set_encoding(crypter.NO_ENCODING)
    ciphertext_bytes = crypter.Encrypt(input)
    assert crypter.Decrypt(ciphertext_bytes) == input

def EncryptAndDecryptCompressed(keyset_path):
    input = 'Secret message'
    crypter = keyczar.Crypter.Read(keyset_path)
    crypter.set_compression(crypter.ZLIB)
    ciphertext_bytes = crypter.Encrypt(input)
    assert crypter.Decrypt(ciphertext_bytes) == input

if __name__ == '__main__':
    if (len(sys.argv) != 2 or not os.path.exists(sys.argv[1])):
        print >> sys.stderr, "Provide a key set path as argument."
        sys.exit(1)
    EncryptAndDecrypt(sys.argv[1])
    EncryptAndDecryptBytes(sys.argv[1])
    EncryptAndDecryptCompressed(sys.argv[1])

File: cpp/src/examples/basic_encrypt.cc
// Encrypts and decrypts a short message. Uses raw encoding, base64w
// encoding and zlib compression.
//
// Example: ./basic_encrypt ~/my-aes
//
#include <cassert>
#include <iostream>
#include <string>

#include <keyczar/keyczar.h>

void EncryptAndDecrypt(const std::string& location) {
  keyczar::Keyczar* crypter = keyczar::Crypter::Read(location);
  if (!crypter)
    return;

  std::string input = "Secret message";
  std::string ciphertext;
  std::cout << "Plaintext: " << input << std::endl;

  bool result = crypter->Encrypt(input, &ciphertext);
  if (result) {
    std::cout << "Ciphertext (base64w): " << ciphertext << std::endl;
    std::string decrypted_input;
    bool result = crypter->Decrypt(ciphertext, &decrypted_input);
    if (result)
      assert(input == decrypted_input);
  }
  delete crypter;
}

void EncryptAndDecryptBytes(const std::string& location) {
  keyczar::Keyczar* crypter = keyczar::Crypter::Read(location);
  if (!crypter)
    return;

  crypter->set_encoding(keyczar::Keyczar::NO_ENCODING);
  assert(crypter->encoding() == keyczar::Keyczar::NO_ENCODING);

  std::string input = "Secret message";
  std::string ciphertext;
  bool result = crypter->Encrypt(input, &ciphertext);
  if (result) {
    std::string decrypted_input;
    bool result = crypter->Decrypt(ciphertext, &decrypted_input);
    if (result)
      assert(input == decrypted_input);
  }
  delete crypter;
}

void EncryptAndDecryptCompressed(const std::string& location) {
  keyczar::Keyczar* crypter = keyczar::Crypter::Read(location);
  if (!crypter)
    return;

  crypter->set_compression(keyczar::Keyczar::ZLIB);
  assert(crypter->compression() == keyczar::Keyczar::ZLIB);

  std::string input = "Secret message";
  std::string ciphertext;
  bool result = crypter->Encrypt(input, &ciphertext);
  if (result) {
    std::string decrypted_input;
    bool result = crypter->Decrypt(ciphertext, &decrypted_input);
    if (result)
      assert(input == decrypted_input);
  }
  delete crypter;
}

int main(int argc, char** argv) {
  if (argc != 2) {
    std::cout << "An absolute key set location must be provided as argument"
              << std::endl;
    return 1;  // error
  }

  // The first argument must represent the keyset's location
  const std::string location(argv[1]);

  EncryptAndDecrypt(location);
  EncryptAndDecryptBytes(location);
  EncryptAndDecryptCompressed(location);

  return 0;
}
